Classes/EventStore/Storage/Doctrine/DoctrineEventStorage.php

```diff
@@ -123,28 +125,40 @@ public function commit(StreamName $streamName, WritableEvents $events, int $expe
         $retryWaitInterval = 0.005;
         $maxRetryAttempts = 8;
         $retryAttempt = 0;
+
+        $rollback = function (Throwable $previous) use (&$retryAttempt, $maxRetryAttempts, &$retryWaitInterval) {
+            if ($retryAttempt >= $maxRetryAttempts) {
+                $this->connection->rollBack();
+                throw new ConcurrencyException(sprintf('Failed after %d retry attempts', $retryAttempt), 1573817175, $previous);
+            }
+            usleep((int)($retryWaitInterval * 1E6));
+            $retryAttempt ++;
+            $retryWaitInterval *= 2;
+            $this->connection->rollBack();
+        };
+
         while (true) {
             $this->reconnectDatabaseConnection();
             if ($this->connection->getTransactionNestingLevel() > 0) {
                 throw new \RuntimeException('A transaction is active already, can\'t commit events!', 1547829131);
             }
             $this->connection->beginTransaction();
             try {
+                # `SELECT FOR UPDATE` write-locks affected data, `SELECT MAX() FOR UPDATE` write-locks the table.
+                # This should be "LOCK TABLE" but "LOCK TABLE" doesn't play nicely with transactions.
+                $this->connection->query('SELECT MAX(sequencenumber) FROM ' . $this->eventTableName . ' FOR UPDATE')->fetchAll();
```

Comment:
Are you sure this does what you expect? https://stackoverflow.com/questions/6593360/select-for-update-and-max suggests otherwise. Also, this means that the table is locked for reading, so this would stall projections, no?